src/Microsoft.AspNet.Razor/TagHelpers/TagHelperDescriptorMatcher.cs

```diff
@@ -0,0 +1,84 @@
+        /// <paramref name="name"/>. <c>null</c> if the attribute is not bound.
+        /// </returns>
+        /// <remarks>
+        /// This method cannot use Reflection to generally determine the most specific <see cref="Type"/> when
+        /// multiple tag helpers bind the attribute. However when it can determine an attribute is bound to a
+        /// <see cref="string"/> property or dictionary value, it always returns <c>"System.String"</c>. Otherwise it
+        /// returns the full <see cref="Type"/> name for the first binding it finds, if any.
+        /// </remarks>
+        public static string GetPropertyType(
+            [NotNull] string name,
+            [NotNull] IEnumerable<TagHelperDescriptor> descriptors)
+        {
+            // Find first TagHelperAttributeDescriptor matching this name.
+            var firstBoundAttribute = descriptors
+                .SelectMany(descriptor => descriptor.Attributes)
+                .FirstOrDefault(attribute => string.Equals(attribute.Name, name, StringComparison.OrdinalIgnoreCase));
+            if (firstBoundAttribute == null)
+            {
+                return null;
+            }
+
+            if (!IsBoundNonStringAttribute(name, descriptors))
+            {
+                // Must have been a string attribute since we know it is bound. Use that type.
+                return typeof(string).FullName;
+            }
+
+            // Attribute is bound but never to a string property.
+            return firstBoundAttribute.TypeName;
+        }
+
+        /// <summary>
+        /// Determines whether an attribute named <paramref name="name"/> is bound exclusively to
+        /// non-<see cref="string"/> tag helper properties.
+        /// </summary>
+        /// <param name="name">Name of the attribute.</param>
+        /// <param name="descriptors">
+        /// Collection of <see cref="TagHelperDescriptor"/>s targeting the containing element in the Razor source.
+        /// </param>
+        /// <returns>
+        /// <c>true</c> if the attribute named <paramref name="name"/> is bound and no associated property has
+        /// <see cref="Type"/> <see cref="string"/>. <c>false</c> otherwise e.g. if the property is not bound.
+        /// </returns>
+        public static bool IsBoundNonStringAttribute(
+            [NotNull] string name,
+            [NotNull] IEnumerable<TagHelperDescriptor> descriptors)
+        {
+            // Find all TagHelperAttributeDescriptor's matching this name.
+            var boundAttributes = descriptors
+                .SelectMany(descriptor => descriptor.Attributes)
+                .Where(attribute => string.Equals(attribute.Name, name, StringComparison.OrdinalIgnoreCase));
+
+            // Check if any matching TagHelperAttributeDescriptor requires a string value.
+            var isStringValue = boundAttributes.Any(attribute => attribute.IsStringProperty);
+
+            return boundAttributes.Any() && !isStringValue;
```

Comment:
No longer relevant but FYI that would not work.  `isStringValue` was `false` in that case but the method was supposed to return `true` then.